### ISSUE TITLE
Changing series' root folder is unable to move the files between drives in mergerfs pool.

### ISSUE THREAD

**Issue by SugarDroid** [Sonarr/Sonarr#4185]
**Describe the bug**
Changing series' root folder is unable to move the files between drives in mergerfs pool.

**Logs**
```
[v3.0.4.1033] System.IO.IOException: Source and destination are not on the same device
  at System.IO.Directory.Move (System.String sourceDirName, System.String destDirName) [0x000dd] in <254335e8c4aa42e3923a8ba0d5ce8650>:0 
  at NzbDrone.Common.Disk.DiskProviderBase.MoveFolder (System.String source, System.String destination) [0x00088] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Common\Disk\DiskProviderBase.cs:263 
  at NzbDrone.Common.Disk.DiskTransferService.TransferFolder (System.String sourcePath, System.String targetPath, NzbDrone.Common.Disk.TransferMode mode) [0x0025e] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Common\Disk\DiskTransferService.cs:88 
  at NzbDrone.Core.Tv.MoveSeriesService.MoveSingleSeries (NzbDrone.Core.Tv.Series series, System.String sourcePath, System.String destinationPath, System.Nullable`1[T] index, System.Nullable`1[T] total) [0x000db] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Core\Tv\MoveSeriesService.cs:59 
```

**System Information**

- Version 3.0.4.1033
- Package Version 3.0.4.1033-ls14 by linuxserver.io
- Mono Version 5.20.1.34

**Additional context**
A move should be move not link + delete, but at least have to fallback if link not possible.

**Comment by bakerboy448** [comment]
xref https://github.com/Radarr/Radarr/issues/5558

**Comment by SugarDroid** [comment]
xref https://github.com/lidarr/Lidarr/issues/1827

**Comment by bakerboy448** [comment]
@SugarDroid  can you grab the trace logs for sonarr specifically and try to get the stack trace?

**Comment by SugarDroid** [comment]
@bakerboy448 the one in OP is not enough, need a duplicate of it? :)

**Comment by bakerboy448** [comment]
@SugarDroid what's your mergerfs settings  

@Taloth is pretty sure it's  mergerfs settings that are preventing renaming the folder.  it may be able to handle it in the arrs, but it is also easy enough to change mergerfs settings if that's the fix :)

**Comment by SugarDroid** [comment]
@bakerboy448 it has nothing to do with my mergerfs settings, but here you go:
```
/media/disk*    /mnt/share  fuse.mergerfs defaults,allow_other,direct_io,dropcacheonclose=true,use_ino,category.create=epmfs,minfreespace=5G    0 0
```
There are 3 issues I opened in tree projects related to the same bug, which happens due bad design. Never mentioned "rename" once neither, so really don't know where the "renaming the folder" idea came from.

Source:
```
/mnt/share$ xattr -p user.mergerfs.allpaths share/media/tv | tr '\0' '\n'
/media/disk3/share/media/tv
/media/disk4/share/media/tv
/media/disk5/share/media/tv
```

Target:
```
/mnt/share$ xattr -p user.mergerfs.allpaths share/media/Tv | tr '\0' '\n'
/media/disk2/share/media/Tv

```
This is what *arrs try to do, what of course will _ALWAYS_ fail anytime it have to face different file systems between source and target:
```
/mnt/share/share/media$ mkdir Tv/test\ folder && ln -v tv/test\ folder/poster.jpg Tv/test\ folder/poster.jpg
ln: failed to create hard link 'Tv/test folder/poster.jpg' => 'tv/test folder/poster.jpg': Invalid cross-device link

```
Sonarr/Lidarr says:  System.IO.IOException: Source and destination are not on the same device
Radarr says: System.IO.IOException: Invalid cross-device link

This is what *arrs have to do for proper operation, what will _NEVER_ fail because move will always fallback to copy+delete anytime if Atomic-Move is not possible:
```
/mnt/share/share/media$ mv -v tv/test\ folder/ Tv/
created directory 'Tv/test folder'
copied 'tv/test folder/poster.jpg' -> 'Tv/test folder/poster.jpg'
copied 'tv/test folder/tvshow.nfo' -> 'Tv/test folder/tvshow.nfo'
removed 'tv/test folder/poster.jpg'
removed 'tv/test folder/tvshow.nfo'
removed directory 'tv/test folder/'

```
And if you guys came up with it, let's see something about renaming too:
```
/mnt/share/share/media$ mv -v Tv/test\ folder/ Tv/Test_folder
renamed 'Tv/test folder/' -> 'Tv/Test_folder'
```

So, according to things above explained/proofed I'm pretty sure that @Taloth is wrong. Without any offense, but stop thinking in a way like "if I use RAID then so does everyone else too". It's false, not everyone use RAID, mergerfs is _not_ RAID. I don't know shit about C# -maybe it's its dumbness-, but I really don't get where the "using link+delete instead move" idea came from, and why anyone thought it's a good idea...

**Comment by Taloth** [comment]
From your original post: `System.IO.Directory.Move (System.String sourceDirName, System.String destDirName)`.

It's telling the filesystem to move the directory (so basically a 'rename'), why can't it? That's normal for between regular filesystems coz you can't move folders cross fs, but why can't mergerfs? I'm no mergerfs expert, but I recall this being related to path preservation settings. You can read more about this here https://github.com/trapexit/mergerfs#path-preservation
and https://github.com/trapexit/mergerfs#rename--link (note them mentioning EXDEV, which is exactly what you're getting)

> what will NEVER fail because move will always fallback to copy+delete anytime if Atomic-Move is not possible:

`mv` is a high level process that tackles a lot of variations including EXDEV. You cannot compare the two.
Also notice how `mv` is copying the files over? This means sonarr would have to _copy_ your media files. This hides the fact that if you configure mergerfs differently, it'll be able to do it instantly.


TL;DR: change the policy to non-path preserving one. Otherwise you'd force apps to copy files across drives for no reason at all.

**Comment by SugarDroid** [comment]
>but why can't mergerfs?

Mergerfs _can_ do it as I showed in the examples, *arr is the one here which is unable to do because not working properly. Every commands ran within mergerfs pool, if `mv` can do it -because it's handled properly- then *arr also have to do it but can't.

> and https://github.com/trapexit/mergerfs#rename--link (note them mentioning EXDEV, which is exactly what you're getting)

You forgot to read the last part which says: "or contacting the author of the offending software and requesting that EXDEV (cross device / improper link) **be properly handled**." It's not mergerfs' fault that it's handled wrong but *arr's.

> Also notice how `mv` is copying the files over? This means sonarr would have to _copy_ your media files.

I'm pretty fine with it because I expect it to be copied between drives, and `mv` do its job perfectly, does fallback to copy+delete if Atomic-Move is not possible. So does *arr have to fallback to copy+delete, but it fails to do its job because -according to @trapexit's words- things are not properly handled.

>This hides the fact that if you configure mergerfs differently...

If i set my mergerfs differently then it won't work anymore in the way I -or many others- _want_, but in a way how _you think_ is good for me.

> TL;DR: change the policy to non-path preserving one. Otherwise you'd force apps to copy files across drives for no reason at all.

I -and pretty sure a lot of others, and this is why the option exist at all- have my reason -which is on the other hand totally irrelevant what they are- to use path preserving. What's your reason -other than laziness- to want to tell me how I have to use my system, just because someone is lazy to fix the faulty working app?

What if -for example- `/media/diskX` is a NFS share from a different server, or an rclone share from any cloud?

Sorry to tell you, this is not solution for anything but a workaround to cover simple laziness. Who would you point to if with this couldn't be possible to workaround *arr's fault, Microsoft or Linus Torvalds?

(I feel your tl;dr here would be something like: install this or that app to do *arr's job.)

Lets get an example: 
You lives in X street of X city. One day you think you would move to Y street in Y city. But the truck driver tells you to keep your car in your old garage in X city, because he forgot to bring his trailer for your car, and just write the new address into documents, and tells it'll be good for you. I'm pretty sure you would be very happy, right?

(This attitude reminds me of Gnome 3's developers who tried to tell people that asking "are you sure to delete x file" before it's moved into trashcan is not enough. So they implemented a new step as "hold down alt key too to be able to delete it at all before we ask you, are you sure..." before it's moved to trashcan and told it will be good for you because I said so...)

TL;DR: Just make things properly handled. If `System.IO.IOException: Source and destination are not on the same device` is the answer for the move command, then fallback to copy+delete, and let the job done -as it should have been from the start- instead failing the process.

**Comment by Taloth** [comment]
> You forgot to read the last part 

Of course I read it. 
 
> What if -for example- `/media/diskX` is a NFS share from a different server, or an rclone share from any cloud?

That's a valid point. But in that case wouldn't you configure mergerfs to either operate as a layered filesystem, with the NFS share being RO/NC and the local drive as RW?
But I do wonder how it interacts with the usual local->cache->read-only clouddrive that some people have. Afaik it's commonly done with a cache mount and an external script to upload async to the cloud (so cloud:RO/NC, cache:RW). Probably with a single local drive.

Please understand that we get frequent support questions regarding file copies vs moves in general, particularly in docker. In general, people don't want unnecessary copies.
Not everyone is a power user. There are plenty of people using mergerfs to simply 'merge' disks, path-preservation forces unnecessary file copies and prevents hardlinks/links in case of torrent seeding. Everyone I've seen so far, turned off path-preservation because that matched their particular use case (just a bunch of disk, without RAID).   
I'm not saying that no other use case exists, we're having this conversation after all, but it's generally the use case we encounter.
In fact, we occasionally recommend mergerfs in this scenario since it's better than having 10 separate mounts and juggling between them.

I'm open to discuss the use cases further. But I like to hear why you don't want it to make the same folder in multiple branches, and secondly how you deal with one drive being full and content being added to it.
Also, do you use torrents on these drives?

> Sorry to tell you, this is not solution for anything but a workaround to cover simple laziness. Who would you point to if with this couldn't be possible to workaround *arr's fault, Microsoft or Linus Torvalds?

Don't do that, it's not constructive. You've used laziness several times. We've had to work around oddities in various filesystems several times, including CIFS copying files partially. This isn't 'just call mv'.
Don't assume laziness or malice, it's unbecoming and for us demotivating.
Explain your particular use case, and ask for the rational behind past decisions, but don't start with that ridiculous blaming game.

> You lives in X street of X city. One day you think you would move to Y street in Y city. But the truck driver tells you to keep your car in your old garage in X city, because he forgot to bring his trailer for your car, and just write the new address into documents, and tells it'll be good for you. I'm pretty sure you would be very happy, right?

Gotta admit, that's a fun analogy.
Here's another one: A car owner has two garages next to his house. He decided that the left garage is for forward-facing cars only, and the right garage is for rear-facing cars only. Now he wants his car to be turned around. Sadly the only way to move the vintage car between garages is to take it apart and move it piece by piece. So some poor sod now has to take the car apart, move the parts one by one to the other garage and reassemble it. So the guy walks up to the owner and says: "Are you sure? It'll be way cheaper to just remove the restriction."   
"No", says the owner, "because this is the way I like it."

**Comment by trapexit** [comment]
I use Sonarr but don't use path preservation so I've not run into any issues but if I understand the thread correctly it sounds like an ask here is to add a 3rd option?

- move
- move + copy on EXDEV
- copy

A move + copy fallback is certainly useful for those who are explicitly managing their directory structures. It is a rare'ish usecase even if it is the default policy. Now that the 'msp' policies exist it is less critical for many.

**Comment by Taloth** [comment]
The ask is for Sonarr to handle nr 2 (move + copy on EXDEV) for directory moves. Mono does it for files internally (it does rename/link+unlink/copy+unlink), but folder moves don't actually exist in the framework, only folder renames so we have to tackle that ourselves.
It makes sense from a filesystem perspective but additional copy operations not what most end users want, especially with torrents. And the reason why so many ppl either have path-preservation off or ignorepponrename enabled, so it preserves path only if possible without copy.
That's the reason why I asked SugarDroid about his usecase in more detail, especially since the example appears to be a reproduction scenario.

**Comment by SugarDroid** [comment]
There was no offense intended in any way, but what could anyone think when the answer for someone's problem is that: "The easiest thing to do is nothing and I'm inclined to do that." It was not you, I get it, but -imho- this is even more destructive to positive thinking when time comes to ask something.

I keep files separated instead a mixed mess, this is why -for example- movie files -and everything related to that movie- are within their own sub directory and not in a big pile of chaos flat directory. (I hear you, there are people who loves that chaos.)

I've a dedicated drive for downloads only. When download is done then files are copied to their own places, movies to one drive, series to another drive, music again to another drive. Every drives has only same type of content. Removing unwanted audio tracks would break link instantly anyway, so...  Torrents -and their data- are deleted after a week of seeding.

**Comment by Taloth** [comment]
So you have like 1 drive for downloads, 1 for movies and 3 for tv. So you're mainly using mergerfs to manage overflow, but are still subdividing your drives. It does feel like a bit micromanagey and you'd end up with more micromanagement when disks full up, but fair enough.   
Your original example showed an attempt to move a series from one of the 3 drives that had the tv branch, onto a new drive. Which appears weird, and probably not what you originally wanted. If you moved every series you have, those 3 drives would be empty and that disk2 full. In what scenario did you _actually_ run into the problem?

As for torrent, so you're not running into the problem that many others are. I gotta say that your usecase is fairly different than the majority.

I'll look into it further.

**Comment by trapexit** [comment]
> It does feel like a bit micromanagey and you'd end up with more micromanagement when disks full up

Yeah. *Most* people don't really need or care about where stuff lands but like the idea till they run into the micromanagement. That's why I created the `msp` set of policies to give a middle of the road strategy.

**Comment by SugarDroid** [comment]
> It does feel like a bit micromanagey and you'd end up with more micromanagement when disks full up, but fair enough.

I don't see much, if any difference between systems related to disk/pool/tank -call it whatever- being full. Let's see few -I promise this time won't be any car in them :smile:- examples:

- One guy has a FreeNAS with 6 disks in RAID, which with time will be full. What can he do? There are only 2 options for him, delete or build a new RAID with 6 new disks.
- Another guy has an unRAID system with 6 disks, which is a mixed mess because he don't care, but with time all disks will be full, regardless non-path preserving. What can he do? There are only 2 options for him too, delete or put at least 1 new disk into his system. 
- And there's a guy -like me, for sure i'm not alone- with a simple Linux system -without shiny WebUI- on bare or brand metal with 6 disks, which holds content separated, instead a mixed mess. With time some disk will be full. What can this guy do? He also has 2 options, delete or put in a new drive and keep his files still separated.

At the end of the "day" all 3 guys will run into micromanagement, as you call it. The only possible difference -beside the price of 6 disks at once for RAID- between the 3 scenarios is the time factor. When... will... it... happen? I don't play in the big boys' league, so 1 disk is enough for quite long time.

> Your original example showed an attempt to move a series from one of the 3 drives that had the tv branch, onto a new drive. Which appears weird, and probably not what you originally wanted. If you moved every series you have, those 3 drives would be empty and that disk2 full. In what scenario did you _actually_ run into the problem?

Is there any weird in that, that someone try to migrate from smaller disks to a bigger, which can hold the content of all and still has free space left for new files? :confused: 

> till they run into the micromanagement. 

Look at the examples above. :smile:

**Comment by Taloth** [comment]
The micromanagey part is that you'd have to delete or add a new drive when your 3 tv drives are full, as opposed to just have it drop the data on whatever drive is still available. So despite having a ton of diskspace free in the pool, you'd get freespace errors in apps. There are several options in mergerfs to deal with the overflow automatically, but instead you've opted for the manual approach.... hence micromanagement. As I said before, doing it like that is fine, but it's definitely not standard.
I suspect the default policy is epmfs rather than mfs, to err on the side of caution since going from mfs to epmfs requires manual work to fix up the directories. Perhaps it's better to have no default at all and force the user to make an educated decision, but that's not the topic of this discussion.

> Is there any weird in that, that someone try to migrate from smaller disks to a bigger, which can hold the content of all and still 
has free space left for new files? 😕

Yes it's weird, because you wouldn't have sonarr do that, you'd drop in the new drive and move the files on the actual drives, thus preserving the exact same paths in mergerfs. Rather than moving from tv to Tv.
So is that what you did? Attempting to migrate from a smaller disk to a bigger one? What's the _actual_ problem you ran into?
If this was your actual operation, then you created 'Tv' on purpose to force the move to the other drive, likely to rename it back to 'tv' later.

**Comment by SugarDroid** [comment]
Using a media manager to manage media is... weird. Gotcha! :boom:

**Comment by Taloth** [comment]
This entire discussion has been about storage management, not media management.

I'm going to assume that's a yes on the questions I asked regarding the rename to Tv and then tv.

This is like using a hammer as a screwdriver. Yeah it'll work (and yes I still planning on making sure it'll work, assuming I can do so for both linux and windows)... but why did you do it like that? Add the drive, move stuff from one drive to the other and Sonarr wouldn't be the wiser, it'd still be ./tv, nothing would've changed from the apps perspective. You would've moved the stored files from one mergerfs branch to the disk2/tv branch one and it'd all have been fine.

My point being that the way you used it is rare, not just the strict mergerfs policy with no exceptions, but also the way you attempted to move the files around with the Tv detour. It's not an approach that a ton of people will make, quite the opposite in fact. You'll likely be both the first and last person to ever run into this issue.

But I think the entire usecase is apparent to me now, thank you for the responses you've given. If I manage to fix it I'll link the commit to this github issue.
Regardless, have a wonderful new year.

**Comment by Phasip** [comment]
Needed to chime in here as I also experienced this but without mergerfs. The generic description may be that Sonarr fails to move files when it doesn't understand that two files are on separate filesystems.
For me the error appears when running Sonarr in a TrueNAS jail with multiple mountpoints in the jail, unfortunately none of the mergerfs features can be used in this case.

IMHO the common user experience is to do a full copy if a link can't be done, possibly with a warning in the logs.

**Comment by kcgthb** [comment]
And the same error happens when trying to delete something, while having the recycle bin on a different device: the move fails, and files are not deleted.
